### PR TITLE
Pull in changes from demo ie end of april

### PR DESCRIPTION
**Pull request by pnkfelix**
I'm not sure if @bryangarza wants to review these, or if I should attempt to factor out something smaller, or if we should just commit to these changes and get pushing forward. I wanted to at least checkpoint the state *before* the commit that changes to the new+improved client-request protocol for `&mut`...

### CODE REVIEW

File: krabcake/rs_hello/src/lib.rs [hunk 1]
```diff
@@ -79,3 +111,939 @@ fn rust_eh_personality() {
...
+    if let Some(event) = STACKED_BORROW_EVENT {
+        before();
+        print_stacked_borrow_event(event);
+        after();
+    }
+}
+
+#[derive(Copy, Clone)]
+enum SbEventKind {
+    BorrowMut = 0x2000,
```
Comment (bryangarza) [inline krabcake/rs_hello/src/lib.rs:201]: same comment as above; where does this number come from?

File: krabcake/rs_hello/src/vex_ir.rs [hunk 2]
```diff
@@ -0,0 +1,2161 @@
+#![allow(non_camel_case_types)]
+
+#[repr(i64)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum IROp {
```
Comment (bryangarza) [inline krabcake/rs_hello/src/vex_ir.rs:5]: side note but at some point we'll want to either make our own libvex crate, or see if one of the existing ones fit our needs

File: krabcake/kc_main.c [hunk 3]
```diff
@@ -113,14 +113,790 @@ static void kc_post_clo_init(void)
...
+      default: tl_assert(0);
+   }
+   t = newTemp(sbOut, ty, k);
+   assign(cat, sbOut, t, e);
+   return mkexpr(t);
+}
+
+typedef enum ExprContext {
+   /* These are the recursive occcurences of IExpr in the grammar of its definition */
+   Expr_GetI_Index = 0x1F00,
```
Comment (bryangarza) [inline krabcake/kc_main.c:234]: where does this number come from? might be good to leave a comment saying what other file it comes from, if there is one (that way it's easy to find later once we forget)

File: krabcake/kc_main.c [hunk 4]
```diff
@@ -113,14 +113,790 @@ static void kc_post_clo_init(void)
...
+                                  IRStmt *st,
+                                  IRExpr* shadow_data )
+{
+   // `t<tmp> = <data>` assigns value to an (SSA) temporary.
+   tl_assert(st->tag == Ist_WrTmp);
+   IRTypeEnv* tyenv = sbOut->tyenv;
+   IRTemp  lhs_tmp = st->Ist.WrTmp.tmp;
+   IRExpr* data = st->Ist.WrTmp.data;
+   IRType  type = typeOfIRExpr(tyenv, data);
+   /* FIXME */
```
Comment (bryangarza) [inline krabcake/kc_main.c:775]: could you add more details this FIXME?

**Review by bryangarza — CHANGES_REQUESTED** [review]

File: krabcake/kc_main.c (hunk 3 above)
Comment (pnkfelix) [reply krabcake/kc_main.c:234]: Oh, its not coming from another file. That's just so I have a better chance of recognizing it in a debugger.

Comment (pnkfelix) [reply krabcake/kc_main.c:234]: Well, okay, *this* number is not coming from another file.

Comment (pnkfelix) [reply krabcake/kc_main.c:234]: The *sequence* of numbers is coming from me scanning over the occurrences of `=0x` in https://github.com/pnkfelix/krabcake-vg/blob/krabcake-vg/VEX/pub/libvex_ir.h; the last one in the file is this:

https://github.com/pnkfelix/krabcake-vg/blob/45a8f67429aa65395176fbab8e721c46f7c40cc1/VEX/pub/libvex_ir.h#L2873

so I took that, and then bumped the `E` to an `F`

I can write a comment explaining this.

File: krabcake/rs_hello/src/vex_ir.rs (hunk 2 above)
Comment (pnkfelix) [reply krabcake/rs_hello/src/vex_ir.rs:5]: (We still need to, at least for now, compile `rs_hello` to a staticlib. But that should not stop us from pulling in another crate that has such definitions, you are right about that.)

File: krabcake/rs_hello/src/lib.rs (hunk 1 above)
Comment (pnkfelix) [reply krabcake/rs_hello/src/lib.rs:201]: You are right, **this** number definitely is shared among multiple files and needs proper referencing.

(or I need to figure out how to pull in the other crate that you made for this here.)

File: krabcake/kc_main.c (hunk 4 above)
Comment (pnkfelix) [reply krabcake/kc_main.c:775]: Heh. There's a lot of FIXME's, and unfortunately I didn't record my thoughts on all of them.

I *think* this fixme is about trying to decide whether the rs_trace_wrtmp routine that we call out to here needs to have a copy of the evaluated `data`, or if its sufficient for it to *solely* observe the `lhs_tmp` and the `shadow_data`...

Comment (pnkfelix) [reply krabcake/kc_main.c:775]: anyway, I'll go over the file and see if I can fill in the missing blanks there.

**Review by bryangarza — APPROVED** [review]

**Pull request merged**